WIP - BP-14 Relax Durability - patch 2 - force() API for FORCE_DEFERRED_ON_JOURNAL LedgerType

eolivelli opened the pull request:
This patch introduce the new force() API which lets Ledgers with LedgerType FORCE_DEFERRED_ON_JOURNAL to force the fsync on journal.
This way the LastAddConfirmed can advance on the client side.

Contents:
- new force() API
- protocol changes (ForceRequest/ForceResponse)
- test cases for client-side API using Mockito
- test cases for Journal/Bookie (still partial)
- test cases for FORCE_DEFERRED_ON_JOURNAL (full round-trip client and bookie)

This patch is marked as Work-In-Progress because I have found a problem during recovery in case of bookie restart. The bug was hidden in my first implementation by implementations in the next patch about "piggybacking LastAddSynced on AddResponse".

For review please take into account second commit https://github.com/apache/bookkeeper/pull/671/commits/3f0f415acbb322f6a05ff41af16b434000616423

eolivelli commented:
@sijie @ivankelly 
Actually the problem is about the recovery of the "LastAddSynced" cursor upon bookie restart.
If the bookie restarts and there is no journal file to process (because a checkpoint happened for instance and all data went to LedgerStorage), scanning the journal will be no-op and so the local position of LastAddSynced for a given ledger is lost.
I tried to use the LastAddConfirmed stored on LedgerStorage and if we piggy-back the lastLastSynced on AddResponse this works partially as the client is able to advance LAC and this LAC gets stored on the bookie. This implementation was also buggy and not correct because it is possible to lose LastSync if no piggy back occours (such piggy back is only an non deterministic optimization).

In the very first implementation (some months ago) on the ForceRequest the client was sending the LastAddPushed entryId to the bookie, supposing that the bookie would answer with that value because if the bookies forces sync on the journal the client could assume that every entry before LastAddPushed would have been persisted. This was not correct too I know (see WriteAdvHandle or out of order responses).

I am thinking about asking to the LedgerStorage the max(id) of the entries stored locally on the bookie.

Another option would be to write on the journal the max(id) of entries written to the disk while flushing the meta 'force' entry. I tried an implementation but it would be very tricky and I did not get to a solution actually

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/bookkeeper-precommit-pullrequest-docker/156/

sijie commented:
@eolivelli I am not sure why do you need to play with journal here. I think I have proposed a solution in the design doc before. Let me try to write a detailed one here:

1) you can have the `LastAddSynced tracker (or cursor) per ledger. it can just simply associated with ledger index.
2) on a file info reloaded, you can reconstruct the cursor: a) you know the `LastAddConfirmed` for a ledger b) read the index file starting from `LastAddConfirmed`, you will know what entries after `LastAddConfirmed` has been persisted by this bookie, stop at a non existed entry. then it is `LastAddSynced`.  
3) cursor tracking are all in memory only, you don't need to write back anything.
4) cursor is updated: a) when an entry is persisted by journal; b) when `LastAddConfirmed` is added. for example, cursor is [las=10, persisted=12, 13, 15, 16]. if 11 is persisted, las will become 13, because all 11, 12, 13 have been persisted. if `LastAddConfirmed` is updated to 14, that means 14 has already been persisted to its ensemble, it is safe to advance last to 16, because all entries before 14 are already persisted in quorum, 15 and 16 are already persisted locally.
5) cursor can be stored in a more compacted format, rather than a flat tree map.
6) for FORCE_ON_JOURNAL ledger type, there should be no cursor for tracking `LastAddSynced`. we should not add any overhead to existing behaviors.

Hope this explain.

Also, since we are changing `sync()` to `force`, please rename `LastAddSynced` to `LastAddForced` as well, to keep consistency.

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java:
```diff
@@ -772,65 +779,66 @@ private String getMyId() throws UnknownHostException {
 
     void readJournal() throws IOException, BookieException {
         long startTs = MathUtils.now();
-        JournalScanner scanner = new JournalScanner() {
```
Ok will change in order to minimize the diff

eolivelli commented:
@sijie you were right. I have dropped SyncCursor from Journal and moved to Bookie. Using your approach the recovery is feasible and I have a test case which failed and now it is working.

I have to cover a remaining point for this patch to be complete (IMHO).
We have to discard such cursor when the bookie deletes the ledger.
I have two ideas:
1) this cursors are in memory only and populated only during writes. If you restart the bookie data is cleared. It is not a great amount of memory. We can live with this minimal memory leak.
2) Leverage GarbageCollectorThread and notify to the bookie when the ledger is deleted from bookie, this way the bookie can discard the cursor for the ledger

eolivelli commented:
I will rename LastAddSynced to LastForced and maybe SyncCursor to a better name (LedgerForceCursor?)

sijie commented:
@eolivelli 

- I haven't checked your implentation. but I assume the cursor is part of FileInfo. so when a gc thread garbage collects a ledger, it would be gone automatically. I am not sure why you are bothering about gc.

- Sync => LastAddForced, SyncCursor => LastAddForcedCursor ?

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/SyncCursor.java:
```diff
@@ -0,0 +1,113 @@
+public class SyncCursor {
+
+    private long minAddSynced = -1;
+    private final SortedMap<Long, Long> ranges = new TreeMap<>();
```
Please take a look at [RangeSet](https://google.github.io/guava/releases/21.0/api/docs/com/google/common/collect/RangeSet.html). 

It is more efficient than using a tree map here. https://github.com/google/guava/wiki/NewCollectionTypesExplained#rangeset
